File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.cpp
class Solution {
public:
    int search(vector<int>& arr, int target) {
        int len=arr.size(), mid;
        int low=0, high=len-1;

        while(low <= high){
            mid = low + (high - low)/2;

            if(arr[mid] == target)
                return mid;

            // for left half
            else if(arr[low] <= arr[mid]){  // increasing sequence
                if(arr[low] <= target && target < arr[mid])
                    high = mid-1;
                else 
                    low = mid+1;
            }

            // for right half
            else if(arr[mid] <= arr[high]){
                if(arr[mid] < target && target <= arr[high]) // "=" for failing tc [5,1,3], target=3
                    low = mid+1;
                else 
                    high = mid-1;               
            }
        }

        return -1;
    }
};

File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.java
class Solution {
  // TC: O(logn)
  // SC: O(1)
  public int search(int[] nums, int target) {
    int len = nums.length;
    int l = 0, r = len - 1, mid;

    while (l <= r) {
      // mid = l + (r - l)/2;
      mid = (l + r) / 2;

      if (nums[mid] == target)
        return mid;

      else if (nums[l] <= nums[mid]) { // for left half
        if (nums[l] <= target && target <= nums[mid])
          r = mid - 1;
        else
          l = mid + 1;
      }

      else { // right half
        if (nums[mid] <= target && target <= nums[r])
          l = mid + 1;
        else
          r = mid - 1;
      }
    }

    return -1;
  }
}